### PR TITLE
Added saving and loading of chats.

### PR DESCRIPTION
**Pull request by deWaardt**
Hi,

I thought this was a very useful feature. I have implemented a messaging service so the contents of the chatbox are available application wide.

After that, I have implemented a method to save and load chats. The elements are fully responsive and I have tried to adhere to your code standards and styled the elements to match your design language.

I will soon make sure it is compatible with dark mode, but for now please check if you like my implementation.

Total list of changes:

### Added MessageService
MessageService now contains the message history so it is available application wide.
I have modified the chat-container to implement this service.

### Added button to toolbar
This button opens the save/load chat dialog

### Added save and load chat dialog
Dialog contains mat-accordion with saved chats. Allows updating existing chats as well as deleting chats.
Dialog is as responsive as I think is realistic. Dynamically updates max-height to control overflow.
Implements messaging service

image

image

image

image

### CODE REVIEW

**Comment by deWaardt** [comment]
Dark mode is not yet implemented.

**Comment by UnknownEnergy** [comment]
That looks awesome, people already asked for this :)
also very good explanation of the pull request.
let me take a closer look at it

File: src/app/toolbar/toolbar.component.ts [hunk 1]
```diff
@@ -63,4 +64,9 @@ export class ToolbarComponent implements OnInit{
     }
     localStorage.setItem('darkModeEnabled', JSON.stringify(this.darkModeEnabled));
   }
+
+  toggleSavePanel() {
+    this.dialog.open(SavedChatsPanelComponent);
+    //this.showSavedChatsPanel = !this.showSavedChatsPanel;
```
Comment (UnknownEnergy) [inline src/app/toolbar/toolbar.component.ts:70]: this comment can be removed I guess

**Comment by UnknownEnergy** [comment]
image

I also noticed it is difficult / not usable for our phone users. Inside the Chrome browser you can simulate a phone screen and try it there

File: src/app/toolbar/saved-chats-panel/saved-chats-panel.component.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  ngOnInit(): void {
+    if (localStorage.getItem('chatHistory') === null) {
+      this.messageHistory = [];
+    } else {
+      this.messageHistory = JSON.parse(localStorage.getItem('chatHistory'));
+      console.log(JSON.parse(localStorage.getItem('chatHistory')));
+
+      this.containerHeight = this.calculateAccordionHeight();
+    }
+    if (this.messageService.currentChatName.length > 0) {
```
Comment (UnknownEnergy) [inline src/app/toolbar/saved-chats-panel/saved-chats-panel.component.ts:36]: I get an error sometimes here on the currentChatName as it is undefined sometimes. Might need to initalize it with an empty string or something.
image

**Comment by deWaardt** [comment]
> image
> 
> I also noticed it is difficult / not usable for our phone users. Inside the Chrome browser you can simulate a phone screen and try it there

I had the idea of changing the dialog’s layout depending on device. I’ll look at an implementation for this soon!

I’ll take care of your comments as soon as I have time!

File: src/app/toolbar/saved-chats-panel/saved-chats-panel.component.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  currentLoadedChatName: string = "";
+  containerHeight: string = '400px';
+
+
+  ngOnInit(): void {
+    if (localStorage.getItem('chatHistory') === null) {
+      this.messageHistory = [];
+    } else {
+      this.messageHistory = JSON.parse(localStorage.getItem('chatHistory'));
+      console.log(JSON.parse(localStorage.getItem('chatHistory')));
```
Comment (UnknownEnergy) [inline src/app/toolbar/saved-chats-panel/saved-chats-panel.component.ts:32]: also we can remove the console logging for production, good for testing but for production not needed :)

File: src/app/toolbar/saved-chats-panel/saved-chats-panel.component.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    while (index < breakpoints.length && height > breakpoints[index]) {
+      index++;
+    }
+
+    return values[index];
+  }
+
+  @HostListener('window:resize', ['$event'])
+  onResize(event?) {
+    console.log(window.innerHeight);
```
Comment (UnknownEnergy) [inline src/app/toolbar/saved-chats-panel/saved-chats-panel.component.ts:59]: same here

**Comment by UnknownEnergy** [comment]
> > image
> > I also noticed it is difficult / not usable for our phone users. Inside the Chrome browser you can simulate a phone screen and try it there
> 
> I had the idea of changing the dialog’s layout depending on device. I’ll look at an implementation for this soon!
> 
> I’ll take care of your comments as soon as I have time!

no stress take your time yes might make sense to do that as we have a lot of phone users this would be needed for them. Awesome work :+1:  we should add some credits page will do this in the future then we can add people like you who did some good contribution to this open-source project :)

File: src/app/toolbar/toolbar.component.html [hunk 5]
```diff
@@ -12,7 +12,11 @@
     <button aria-label="Info Dialog" class="btn btn-primary" (click)="openInfoDialog()">
       <i class="bi bi-info-circle"></i>
     </button>
+    <button aria-label="Chat history" class="btn btn-primary" (click)="toggleSavePanel()">
```
Comment (UnknownEnergy) [inline src/app/toolbar/toolbar.component.html:15]: another thing please also add an aria-label to every button you added which does not have a text itself as it is required for people which do not use/can use keyboards and use voice instead. Someone asked for this feature some time ago.

**Comment by deWaardt** [comment]
https://streamable.com/2gci4r

I've put in a little effort today to make the dialog more responsive. It's almost working fine on all screen sizes now.
I'm a bit short on time, I'll finish it up somewhere next week 👍

**Comment by UnknownEnergy** [comment]
> https://streamable.com/2gci4r
> 
> I've put in a little effort today to make the dialog more responsive. It's almost working fine on all screen sizes now. I'm a bit short on time, I'll finish it up somewhere next week +1

looks awesome nice work

**Comment by UnknownEnergy** [comment]
hey @deWaardt any update on the feature? :D 
would be cool if we can integrate it.

**Comment by deWaardt** [comment]
Hey!

I’m sorry, it kinda flew under the radar. I have been very sick for a period and forgot about the project under work stress I had to catch up to.

I will have the feature finished soon, I’m gonna continue working on it this week!

**Comment by deWaardt** [comment]
I've mostly finalized it, I'm gonna do the finishing touches tomorrow and then I'll post it so you can review it. My brain is still a little cloudy from being pretty sick at the moment so please double check my pull request 😢

**Comment by UnknownEnergy** [comment]
thank you very much people are already waiting for this in some other feature request it looks awesome :+1: 
Yes no stress make sure to get well soon best wishes

**Comment by deWaardt** [comment]
I'm sorry for the delay, I am mostly better now.

I've updated the PR. The background color in dark mode isn't correct yet, I'll correct that ASAP.
But please check over the code.

My brain is a bit messy still, I tested it thoroughly and everything seems to work well.

**Comment by UnknownEnergy** [comment]
yeah take your time just merged it now and will deploy it soon 
looks good to me

**Pull request merged**